packages/@lwc/engine-core/src/framework/hydration.ts

```diff
@@ -80,14 +78,92 @@ type AttrValidationPredicate = (attrName: string) => boolean;
 
 // flag indicating if the hydration recovered from the DOM mismatch
 let hasMismatch = false;
+
+// Errors that occured during the hydration process
+let hydrationErrors: Array<HydrationError> = [];
+
+class HydrationError {
+    type: string;
+    serverRendered: any;
+    clientExpected: any;
+
+    constructor(type: string, serverRendered?: any, clientExpected?: any) {
+        this.type = type;
+        this.serverRendered = serverRendered;
+        this.clientExpected = clientExpected;
+    }
+}
+
+/*
+    Hydration errors occur before the source node has been fully hydrated,
+    queue them so they can be logged later against the mounted node.
+*/
+function queueHydrationError(type: string, serverRendered?: any, clientExpected?: any) {
+    if (process.env.NODE_ENV !== 'production') {
+        hydrationErrors.push(new HydrationError(type, serverRendered, clientExpected));
+    }
+}
+
+/*
+    Flushes (logs) any queued errors after the source node has been mounted.
+ */
+function flushHydrationErrors(source?: Node | null) {
+    if (process.env.NODE_ENV !== 'production') {
+        hydrationErrors.forEach((error) =>
+            logWarn(
+                `Hydration ${error.type} mismatch on:`,
+                source,
+                `\n- rendered on server:`,
+                error.serverRendered,
+                `\n- expected on client:`,
+                error.clientExpected || source
```

Comment:
The fact that the newlines and string length cause everything to line up is a really nice touch.